File: src/main/java/com/starshootercity/abilities/FireImmunity.java
package com.starshootercity.abilities;

import com.starshootercity.OriginSwapper;
import net.kyori.adventure.key.Key;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageEvent;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class FireImmunity implements VisibleAbility, Listener {
    @EventHandler
    public void onEntityDamageEvent(EntityDamageEvent event) {
        AbilityRegister.runForAbility(event.getEntity(), getKey(), () -> {
            List<EntityDamageEvent.DamageCause> causes = new ArrayList<>() {{
                add(EntityDamageEvent.DamageCause.FIRE);
                add(EntityDamageEvent.DamageCause.FIRE_TICK);
                add(EntityDamageEvent.DamageCause.LAVA);
                add(EntityDamageEvent.DamageCause.HOT_FLOOR);
            }};

            if (causes.contains(event.getCause())) {
                event.setCancelled(true);
            }
        });
    }

    @Override
    public @NotNull Key getKey() {
        return Key.key("origins:fire_immunity");
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getDescription() {
        return OriginSwapper.LineData.makeLineFor("You are immune to all types of fire damage.", OriginSwapper.LineData.LineComponent.LineType.DESCRIPTION);
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getTitle() {
        return OriginSwapper.LineData.makeLineFor("Fire Immunity", OriginSwapper.LineData.LineComponent.LineType.TITLE);
    }
}

File: core/src/main/kotlin/ru/turbovadim/abilities/FireImmunity.kt
package ru.turbovadim.abilities

import ru.turbovadim.OriginSwapper.LineData.Companion.makeLineFor
import ru.turbovadim.OriginSwapper.LineData.LineComponent
import ru.turbovadim.abilities.Ability.AbilityRunner
import net.kyori.adventure.key.Key
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageEvent

class FireImmunity : VisibleAbility, Listener {

    @EventHandler
    fun onEntityDamageEvent(event: EntityDamageEvent) {
        val fireCauses = setOf(
            EntityDamageEvent.DamageCause.FIRE,
            EntityDamageEvent.DamageCause.FIRE_TICK,
            EntityDamageEvent.DamageCause.LAVA,
            EntityDamageEvent.DamageCause.HOT_FLOOR
        )

        if (event.cause in fireCauses) {
            runForAbility(event.entity, AbilityRunner { _ ->
                event.isCancelled = true
            })
        }
    }

    override fun getKey(): Key {
        return Key.key("origins:fire_immunity")
    }

    override val description: MutableList<LineComponent> = makeLineFor(
        "You are immune to all types of fire damage.",
        LineComponent.LineType.DESCRIPTION
    )

    override val title: MutableList<LineComponent> = makeLineFor(
        "Fire Immunity",
        LineComponent.LineType.TITLE
    )
}
